modules/lang-painless/src/main/java/org/elasticsearch/painless/AnalyzerCaster.java

```diff
@@ -722,9 +722,10 @@ public static Cast getLegalCast(Location location, Type actual, Type expected, b
                 break;
         }
 
-        if (actual.sort == Sort.DEF || expected.sort == Sort.DEF ||
-            expected.clazz.isAssignableFrom(actual.clazz) ||
-            explicit && actual.clazz.isAssignableFrom(expected.clazz)) {
+        if (       actual.sort == Sort.DEF
+                || (actual.sort != Sort.VOID && expected.sort == Sort.DEF)
+                || expected.clazz.isAssignableFrom(actual.clazz)
+                || (explicit && actual.clazz.isAssignableFrom(expected.clazz))) {
```

Comment:
No. I just can never remember if `&&` comes before `||` so I always use parentheses to make them clear to me.